op-node/rollup/driver/finalized.go

```diff
@@ -0,0 +1,33 @@
+package driver
+
+import (
+	"context"
+
+	"github.com/ethereum/go-ethereum"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+type finalized struct {
+	derive.L1Fetcher
+	l1Finalized func() eth.L1BlockRef
+}
+
+func NewFinalized(l1Finalized func() eth.L1BlockRef, fetcher derive.L1Fetcher) *finalized {
+	return &finalized{L1Fetcher: fetcher, l1Finalized: l1Finalized}
+}
+
+func (f *finalized) L1BlockRefByNumber(ctx context.Context, num uint64) (eth.L1BlockRef, error) {
+	// Don't apply the finalized wrapper if l1Finalized is empty (as it is during the startup case before the l1State is initialized).
```

Comment:
i don't understand this case...

don't we end with a case where we end up relying on a block that might not be finalized with this condition?

Reply:
This is similar to the code in `confDepth`:

https://github.com/celo-org/optimism/blob/b2769e655678559fda856c869a202be913e39458/op-node/rollup/driver/conf_depth.go#L34-L38

This was added in https://github.com/ethereum-optimism/optimism/pull/3737 and I think the same argument applies here.

Reply:
i guess.. i understand why it would be empty...

what i don't understand if it's ok to then just use whatever block number we're given.... (instead of let's say wait until we have a block we know it's final)

Reply:
This was indeed an issue, I think upstream has it in their code as well, but it matters less there, as the number of confirmation blocks is usually low and sure to be triggered (+ they are ok with reorgs).

The following logs show the implementation where the block is returned when the `status.l1Finalized` is not yet set.

While it isn't set, the L1 block advances with the L1 block, which makes the system accept non-finalized blocks.

```log
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0x0000000000000000000000000000000000000000000000000000000000000000:0 num=31
op-node-1  lvl=warn msg="not set" l1Finalized=0x0000000000000000000000000000000000000000000000000000000000000000:0 num=31
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x54590afe676be0cac737eb1cf7738d3205076562fa8da423aca9586e26e7ae97:30 next=0x3eb9e68eae3d12203eeabe811a2a6c3bbe40f3aca8f02fa3eee177384116e67b:31
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0x0000000000000000000000000000000000000000000000000000000000000000:0 num=32
op-node-1  lvl=warn msg="not set" l1Finalized=0x0000000000000000000000000000000000000000000000000000000000000000:0 num=32
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x3eb9e68eae3d12203eeabe811a2a6c3bbe40f3aca8f02fa3eee177384116e67b:31 next=0x0000000000000000000000000000000000000000000000000000000000000000:0
```

Once `l1FInalized` is set, the node waits until the system catches up.

```log
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0x54590afe676be0cac737eb1cf7738d3205076562fa8da423aca9586e26e7ae97:30 num=32
op-node-1  lvl=warn msg="not finalized" num=32 l1Finalized=0x54590afe676be0cac737eb1cf7738d3205076562fa8da423aca9586e26e7ae97:30
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x3eb9e68eae3d12203eeabe811a2a6c3bbe40f3aca8f02fa3eee177384116e67b:31 next=0x0000000000000000000000000000000000000000000000000000000000000000:0
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0x54590afe676be0cac737eb1cf7738d3205076562fa8da423aca9586e26e7ae97:30 num=32
op-node-1  lvl=warn msg="not finalized" num=32 l1Finalized=0x54590afe676be0cac737eb1cf7738d3205076562fa8da423aca9586e26e7ae97:30
op-node-1  lvl=warn msg="PollBlockChanges" label=safe
op-node-1  lvl=warn msg="PollBlockChanges" label=finalized
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0x69742b037f410a17119decc08ea7ac55718cb5094cd7e95a1e2274025667b6ec:38 num=32
op-node-1  lvl=warn msg="is finalized" num=32 l1Finalized=0x69742b037f410a17119decc08ea7ac55718cb5094cd7e95a1e2274025667b6ec:38
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x3eb9e68eae3d12203eeabe811a2a6c3bbe40f3aca8f02fa3eee177384116e67b:31 next=0x525071112a3e436475c9f9e60712040a3209b1a5d12cd371d5a49f8356fec91d:32
```

What should happen instead is that the the block number is never updated to higher values than the finalized block. By removing the check, this works as expected:

```log
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0x0000000000000000000000000000000000000000000000000000000000000000:0 num=2
op-node-1  lvl=warn msg="not finalized" num=2 l1Finalized=0x0000000000000000000000000000000000000000000000000000000000000000:0
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0xc4bde21356d528c88ff101b2d020cc45ec001ececffc9b0e3c545c96a6e2d2fa:1 next=0x0000000000000000000000000000000000000000000000000000000000000000:0
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0x0000000000000000000000000000000000000000000000000000000000000000:0 num=2
op-node-1  lvl=warn msg="not finalized" num=2 l1Finalized=0x0000000000000000000000000000000000000000000000000000000000000000:0
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=2
op-node-1  lvl=warn msg="is finalized" num=2 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0xc4bde21356d528c88ff101b2d020cc45ec001ececffc9b0e3c545c96a6e2d2fa:1 next=0xfd37afaf9373373653e3fa77db62fbb9344a6a3825759bc87e14e035241d4a37:2
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=3
op-node-1  lvl=warn msg="is finalized" num=3 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0xfd37afaf9373373653e3fa77db62fbb9344a6a3825759bc87e14e035241d4a37:2 next=0x9b6b929ec73386c74048ed43200bb0c062c81a4c0f731f794df362731641951b:3
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=4
op-node-1  lvl=warn msg="is finalized" num=4 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x9b6b929ec73386c74048ed43200bb0c062c81a4c0f731f794df362731641951b:3 next=0x3090146786105ad9eada0c9d9a27fed9e32f1e0312eca28cfa71cffc249f6d53:4
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=5
op-node-1  lvl=warn msg="is finalized" num=5 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x3090146786105ad9eada0c9d9a27fed9e32f1e0312eca28cfa71cffc249f6d53:4 next=0x3a34f8671d2b5da403bc1ae2e247e17a71217af1867af0f22913c7062dc6cb70:5
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=6
op-node-1  lvl=warn msg="is finalized" num=6 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x3a34f8671d2b5da403bc1ae2e247e17a71217af1867af0f22913c7062dc6cb70:5 next=0x6f4866e9cac01c8def4f45c9616fe55d61624deabefe4d859f666d36e5952c16:6
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=7
op-node-1  lvl=warn msg="is finalized" num=7 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x6f4866e9cac01c8def4f45c9616fe55d61624deabefe4d859f666d36e5952c16:6 next=0xeb4391ce24017e010a10d2ff2457a42fa77fb75ed37bd5db5543fec23e6d8a19:7
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=8
op-node-1  lvl=warn msg="is finalized" num=8 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0xeb4391ce24017e010a10d2ff2457a42fa77fb75ed37bd5db5543fec23e6d8a19:7 next=0x3b4f55d0ff6661bc45b0b905fdd9f3f749f7f3e61ff2acd319ff7a57956285dd:8
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=9
op-node-1  lvl=warn msg="is finalized" num=9 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x3b4f55d0ff6661bc45b0b905fdd9f3f749f7f3e61ff2acd319ff7a57956285dd:8 next=0x34dcbff8121c21590a87714414ac7ff031b52bc4323b9030de405aab08b6f11c:9
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=10
op-node-1  lvl=warn msg="is finalized" num=10 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x34dcbff8121c21590a87714414ac7ff031b52bc4323b9030de405aab08b6f11c:9 next=0x9ed81f657faf7af7b622353f4d74d97c1cc7f7d6e2dc4fd6c692340dc69ed218:10
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=11
op-node-1  lvl=warn msg="is finalized" num=11 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x9ed81f657faf7af7b622353f4d74d97c1cc7f7d6e2dc4fd6c692340dc69ed218:10 next=0xd2fcdaf81a4a4c5d91590e8665e2a2f063edc278c44b283b92e774ceca432a65:11
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=12
op-node-1  lvl=warn msg="is finalized" num=12 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0xd2fcdaf81a4a4c5d91590e8665e2a2f063edc278c44b283b92e774ceca432a65:11 next=0x702b905688c2123a1c39fbf589fa902ae719eef4cb7c974fc02bc4b9a1e81e5a:12
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=13
op-node-1  lvl=warn msg="is finalized" num=13 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0x702b905688c2123a1c39fbf589fa902ae719eef4cb7c974fc02bc4b9a1e81e5a:12 next=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="L1BlockRefByNumber" l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 num=14
op-node-1  lvl=warn msg="not finalized" num=14 l1Finalized=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13
op-node-1  lvl=warn msg="AdvanceL1Block" origin=0xa67e5108f5c2ac9f93813d2512b902f023846c925bb07b75af8515621010e626:13 next=0x0000000000000000000000000000000000000000000000000000000000000000:0
```